src/System.Private.CoreLib/shared/System/Runtime/Intrinsics/Vector128`1.cs

```diff
@@ -0,0 +1,76 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using Internal.Runtime.CompilerServices;
+
+namespace System.Runtime.Intrinsics
+{
+    [Intrinsic]
+    [DebuggerDisplay("{DisplayString,nq}")]
+    [DebuggerTypeProxy(typeof(Vector128DebugView<>))]
+    [StructLayout(LayoutKind.Sequential, Size = 16)]
+    public readonly struct Vector128<T> where T : struct
+    {
+        // These fields exist to ensure the alignment is 8, rather than 1.
+        // This also allows the debug view to work https://github.com/dotnet/coreclr/issues/15694)
+        private readonly ulong _00;
+        private readonly ulong _01;
+
+        private unsafe string DisplayString
+        {
+            get
+            {
+                // The IsPrimitive check ends up working for `bool`, `char`, `IntPtr`, and `UIntPtr`
+                // which are not actually supported by any current architecture. This shouldn't be
+                // an issue however and greatly simplifies the check
+
+                if (typeof(T).IsPrimitive)
+                {
+                    var items = new T[16 / Unsafe.SizeOf<T>()];
+                    Unsafe.WriteUnaligned(ref Unsafe.As<T, byte>(ref items[0]), this);
+                    return $"({string.Join(", ", items)})";
+                }
+                else
+                {
+                    return SR.NotSupported_Type;
+                }
+            }
+        }
+
+        public static Vector128<T> Zero => default;
+
+        public Vector128<byte> AsByte() => Unsafe.As<Vector128<T>, Vector128<byte>>(ref Unsafe.AsRef(in this));
+
+        public Vector128<double> AsDouble() => Unsafe.As<Vector128<T>, Vector128<double>>(ref Unsafe.AsRef(in this));
+
+        public Vector128<short> AsInt16() => Unsafe.As<Vector128<T>, Vector128<short>>(ref Unsafe.AsRef(in this));
+
+        public Vector128<int> AsInt32() => Unsafe.As<Vector128<T>, Vector128<int>>(ref Unsafe.AsRef(in this));
+
+        public Vector128<long> AsInt64() => Unsafe.As<Vector128<T>, Vector128<long>>(ref Unsafe.AsRef(in this));
+
+        [CLSCompliant(false)]
+        public Vector128<sbyte> AsSByte() => Unsafe.As<Vector128<T>, Vector128<sbyte>>(ref Unsafe.AsRef(in this));
+
+        public T AsScalar() => Unsafe.As<Vector128<T>, T>(ref Unsafe.AsRef(in this));
```

Comment:
Do we need to handle the case where the user specifies an "unsupported" `T`? That is, we currently only support `byte`, `sbyte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `float`, and `double`. We may also support other types in the future (such as `Half`).

The current code will work for any `T` which is a struct and is less than or equal to the size of the containing vector; but will read past the end of the struct if the user passes in an unsupported type larger than that (say they pass in `MyStruct`).

For the software implementation, should we explicitly have the appropriate check here?

Reply:
If we do want an "appropriate" check, we could either have an explicit check:
```csharp
if (typeof(T) != typeof(byte) && ... && typeof(T) != typeof(double))
{
    throw new PlatformNotSupportedException();
}
```

or we could have a less restrictive but still "correct" check:
```csharp
if (Unsafe.SizeOf<T>() > 8) // Vector64
if (Unsafe.SizeOf<T>() > 16) // Vector128
if (Unsafe.SizeOf<T>() > 32) // Vector256
```

Reply:
I prefer the first approach (explicit check).